ReactAndroid/src/main/java/com/facebook/react/views/scroll/MaintainVisibleScrollPositionHelper.java

```diff
@@ -0,0 +1,186 @@
+import com.facebook.react.bridge.UIManager;
+import com.facebook.react.bridge.UIManagerListener;
+import com.facebook.react.bridge.UiThreadUtil;
+import com.facebook.react.uimanager.UIManagerHelper;
+import com.facebook.react.uimanager.common.ViewUtil;
+import com.facebook.react.views.view.ReactViewGroup;
+import com.facebook.react.views.scroll.ReactScrollViewHelper.HasSmoothScroll;
+
+import java.lang.ref.WeakReference;
+
+/**
+ * Manage state for the maintainVisibleContentPosition prop.
+ *
+ * This uses UIManager to listen to updates and capture position of items before and after layout.
+ */
+public class MaintainVisibleScrollPositionHelper<ScrollViewT extends ViewGroup & HasSmoothScroll> implements UIManagerListener {
+  private final ScrollViewT mScrollView;
+  private final boolean mHorizontal;
+  private @Nullable Config mConfig;
+  private @Nullable WeakReference<View> mFirstVisibleView = null;
+  private @Nullable Rect mPrevFirstVisibleFrame = null;
+  private boolean mListening = false;
+
+  public static class Config {
+    public final int minIndexForVisible;
+    public final @Nullable Integer autoScrollToTopThreshold;
+
+    Config(int minIndexForVisible, @Nullable Integer autoScrollToTopThreshold) {
+      this.minIndexForVisible = minIndexForVisible;
+      this.autoScrollToTopThreshold = autoScrollToTopThreshold;
+    }
+
+    static Config fromReadableMap(ReadableMap value) {
+      int minIndexForVisible = value.getInt("minIndexForVisible");
+      Integer autoScrollToTopThreshold =
+        value.hasKey("autoscrollToTopThreshold")
+          ? value.getInt("autoscrollToTopThreshold")
+          : null;
+      return new Config(minIndexForVisible, autoScrollToTopThreshold);
+    }
+  }
+
+  public MaintainVisibleScrollPositionHelper(ScrollViewT scrollView, boolean horizontal) {
+    mScrollView = scrollView;
+    mHorizontal = horizontal;
+  }
+
+  public void setConfig(@Nullable Config config) {
+    mConfig = config;
+  }
+
+  /**
+   * Start listening to view hierarchy updates. Should be called when this is created.
+   */
+  public void start() {
+    if (mListening) {
+      return;
+    }
+    mListening = true;
+    getUIManagerModule().addUIManagerEventListener(this);
+  }
+
+  /**
+   * Stop listening to view hierarchy updates. Should be called before this is destroyed.
+   */
+  public void stop() {
+    if (!mListening) {
+      return;
+    }
+    mListening = false;
+    getUIManagerModule().removeUIManagerEventListener(this);
+  }
+
+  /**
+   * Update the scroll position of the managed ScrollView. This should be called after layout
+   * has been updated.
+   */
+  public void updateScrollPosition() {
+    if (mConfig == null
+      || mFirstVisibleView == null
+      || mPrevFirstVisibleFrame == null) {
+      return;
+    }
+
+    View firstVisibleView = mFirstVisibleView.get();
+    Rect newFrame = new Rect();
+    firstVisibleView.getHitRect(newFrame);
+
+    if (mHorizontal) {
```

Comment:
Do you have an example of the interface you have in mind?

My thinking when organizing the code like this was to reduce duplication between the 2 implementations while also moving all code related to maintain visible content position out of the scrollview classes to avoid making them too big. The common logic when the variant specific is also very intertwined in some cases, see https://github.com/facebook/react-native/pull/35049/files/c069ae3e39595e7d7e7a3259c6004d1313ad6ba8#diff-b0c8b7c192810637724d9d2ed35db2b8921c45e153ebbda1ef8f34cb53837473R153 for example.